pinot-core/src/main/java/org/apache/pinot/core/query/scheduler/QueryScheduler.java

```diff
@@ -168,6 +169,19 @@ protected byte[] processQueryAndSerialize(ServerQueryRequest queryRequest, Execu
 
       byte[] responseBytes = serializeResponse(queryRequest, instanceResponse);
 
+      Map<String, String> queryOptions = queryRequest.getQueryContext().getQueryOptions();
+      Long maxResponseByteLength = QueryOptionsUtils.getMaxSerializedResponseLengthPerServer(queryOptions);
+      if (maxResponseByteLength != null && responseBytes.length > maxResponseByteLength) {
```

Comment:
Is it better to do this sooner as the serialized response is being built ? There is a method that converts DataTable into toBytes(). The method does incrementally by going over various data structures in the DataTable. Ideally we can intersperse this check within that method at various points ?

On a related note, is it possible to do this check as segment level individual result is being combined by the CombineOperator ?

Reply:
+1. The sooner will kill it the less time and memory we spend on it.

Reply:
I did evaluate that option. But we currently do not have a good mechanism for handling exceptions during serializing and effectively return to the user. I've added a TODO here.